File: Day224/minimumNumberOfTapToOpenToWaterAGarden.go
package main
import (
	"fmt"
	"sort"
)
func minTaps(n int, ranges []int) int {
	arr := make([][]int, n+1)
	for i := range arr {
		arr[i] = make([]int, 2)
		arr[i][0] = max(0, i - ranges[i])
		arr[i][1] = min(n, i + ranges[i])
	}

	sort.Slice(arr, func(i, j int) bool {
		if arr[i][0]==arr[j][0]{
			return arr[j][1]<arr[i][1]
		}else{
			return arr[i][0]<arr[j][0]
		}
	})
	fmt.Println(arr)
	if arr[0][0]>0{
		return -1
	}
	end:=arr[0][1]
	count:=0
	for i:=range arr{
		count++
		if end>=n{
			return count
		}
		newEnd:=end
		for j:=i+1; j<len(arr); j++{
			if arr[j][0]<=end&&arr[j][1]>newEnd{
				i=j
				newEnd = arr[j][1]
			}
		}
		if newEnd==end{
			return -1
		}
		end = newEnd
		i--
	}
	return -1
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func main() {
	arr := []int{3,4,1,1,0,0}
	fmt.Println(minTaps(5,arr))
}

File: Day224/minimumNumberOfTapToOpenToWaterAGarden.java
class Solution {
    public int minTaps(int n, int[] ranges) {
        int arr[][] = new int[n+1][2];
				for (int i=0; i<=n; i++) {
					arr[i][0] = Math.max(0, i - ranges[i]);
					arr[i][1] = Math.min(n, i + ranges[i]);
				}
				Arrays.sort(arr, (a,b)->{
					if (a[0] == b[0]) {
						return b[1] - a[1];
					} else {
						return a[0] - b[0];
					}
				});
				// for (int i=0; i<=n; i++) {
				// 	System.out.println(arr[i][0] +" "+ arr[i][1]);
				// }
				if (arr[0][0] > 0) {
					return -1;
				}
				int end = arr[0][1];
				int count = 0;
				for (int i=0; i<=n; i++) {
					count++;
					if (end >= n) {
						return count;
					}
					int newEnd = end;
					for (int j=i+1; j<n; j++) {
						if (arr[j][0] <= end && arr[j][1]>newEnd) {
							i = j;
							newEnd = arr[j][1];
						}
					}
					if (newEnd == end) {
						return -1;
					}
					end = newEnd;
					i--;
				}
				return -1;
    }
}